include/pybind11/detail/internals.h

```diff
@@ -11,6 +11,19 @@
 
 #include "../pytypes.h"
 
+/// Tracks the `internals` and `type_info` ABI version independent of the main library version.
+///
+/// Some portions of the code use an ABI that is conditional depending on this
+/// version number.  That allows ABI-breaking changes to be "pre-implemented".
+/// Once the default version number is incremented, the conditional logic that
+/// no longer applies can be removed.  Additionally, users that need not
+/// maintain ABI compatibility can increase the version number in order to take
+/// advantage of any functionality/efficiency improvements that depend on the
+/// newer ABI.
+#ifndef PYBIND11_INTERNALS_VERSION
+#define PYBIND11_INTERNALS_VERSION 4
```

Comment:
Is manually increasing this "supported"? The text above seems to imply so, but there are no tests with this increased. I think we clearly need to indicate that this is a "use at your own risk" situation (since we might add more to the new ABI before officially increasing it), or add tests, or both. I like both. :)

Reply:
I added a warning.  As far as tests, I did test locally with both ABI versions.  Adding the tests to CI would potentially double the number of test configurations, though, which may be undesirable.

Reply:
Yes, which is one reason I'm a little hesitant about duplicating things here. I'd be fine with just 1-2 tests for this, though not ideal, still better than having a bunch of completely untested code.

Reply:
@henryiii asked: Is manually increasing this "supported"?

@jbms Is the plan to patch the internals version when we import into the Google codebase?

Reply:
Yeah, I assume for internal use at Google we would increase the version number.  That could be done via a `define` in the BUILD file without having to patch the source code.